File: tooling/architecture/js_unreached_assertions.mjs
import { readFileSync } from "node:fs";
import * as espree from "espree";

const PARSE = { ecmaVersion: "latest", sourceType: "module", loc: true };

const FN = new Set([
    "FunctionExpression",
    "ArrowFunctionExpression",
    "FunctionDeclaration",
]);

// A callback the test body itself drives: its assertions run in the test's own
// straight-line flow, so non-invocation is not a risk.
const DRIVEN_BY_CALLER = new Set([
    "map",
    "forEach",
    "filter",
    "find",
    "findIndex",
    "some",
    "every",
    "flatMap",
    "reduce",
    "sort",
    "then",
    "catch",
    "finally",
    "toThrow",
    "expect",
]);

function walk(node, fn, parent = null) {
    if (!node || typeof node.type !== "string") return;
    fn(node, parent);
    for (const key of Object.keys(node)) {
        if (key === "loc" || key === "range" || key === "parent") continue;
        const child = node[key];
        if (Array.isArray(child)) {
            for (const c of child)
                if (c && typeof c.type === "string") walk(c, fn, node);
        } else if (child && typeof child.type === "string") walk(child, fn, node);
    }
}

const within = (inner, outer) =>
    inner.range[0] >= outer.range[0] && inner.range[1] <= outer.range[1];

function isTestRegistration(node) {
    if (node.type !== "CallExpression") return false;
    const c = node.callee;
    if (c.type === "Identifier") return c.name === "test";
    return (
        c.type === "MemberExpression" &&
        c.object?.type === "Identifier" &&
        c.object.name === "test" &&
        ["tags", "only", "debug", "skip", "todo"].includes(c.property?.name)
    );
}

/** Every identifier the function assigns to or updates — a sentinel candidate. */
function assignedNames(fn) {
    const names = new Set();
    walk(fn, (n) => {
        if (n.type === "AssignmentExpression" && n.left.type === "Identifier")
            names.add(n.left.name);
        if (n.type === "UpdateExpression" && n.argument.type === "Identifier")
            names.add(n.argument.name);
    });
    return names;
}

for (const file of process.argv.slice(2)) {
    const text = readFileSync(file, "utf8");
    let ast;
    try {
        ast = espree.parse(text, { ...PARSE, range: true });
    } catch (error) {
        process.stderr.write(`${file}: ${error.message}\n`);
        process.exitCode = 1;
        continue;
    }
    walk(ast, (node) => {
        if (!isTestRegistration(node)) return;
        const body = node.arguments.find((a) => a && FN.has(a.type));
        if (!body) return;
        const bodyText = text.slice(body.range[0], body.range[1]);
        // The test pins how many assertions must run, or uses the step protocol.
        if (/expect\.assertions\s*\(|verifySteps\s*\(/.test(bodyText)) return;

        // Deferred functions: declared inside the test, not driven by it.
        const deferred = [];
        walk(body, (n, parent) => {
            if (n === body) return;
            const isFn = FN.has(n.type);
            const isMethod = n.type === "MethodDefinition" && n.kind !== "constructor";
            if (!isFn && !isMethod) return;
            if (
                isFn &&
                parent?.type === "CallExpression" &&
                parent.arguments.includes(n)
            ) {
                const callee = parent.callee;
                const name =
                    callee.type === "MemberExpression"
                        ? callee.property?.name
                        : callee.name;
                if (DRIVEN_BY_CALLER.has(name)) return;
            }
            if (isFn && parent?.type === "MethodDefinition") return; // counted as the method
            deferred.push(
                isMethod ? Object.assign(n.value, { _method: n.key?.name }) : n,
            );
        });
        if (!deferred.length) return;

        // Hoot already fails a test that runs no assertion at all, so a test
        // whose ONLY assertions sit in a dead handler is caught today. What is
        // not caught is a dead handler MASKED by an assertion in the test's own
        // flow: one assertion runs, the test is green, and the assertion that
        // names the behaviour never executed. That masking is the defect.
        let masked = false;
        walk(body, (n) => {
            if (masked) return;
            if (n.type !== "CallExpression" || n.callee.type !== "Identifier") return;
            if (n.callee.name !== "expect") return;
            if (deferred.some((d) => within(n, d))) return;
            masked = true;
        });
        if (!masked) return;

        for (const fn of deferred) {
            let asserts = false;
            walk(fn, (n) => {
                if (
                    n.type === "CallExpression" &&
                    n.callee.type === "Identifier" &&
                    n.callee.name === "expect"
                ) {
                    // ignore an expect() nested in a deeper deferred function
                    const deeper = deferred.some(
                        (d) => d !== fn && within(d, fn) && within(n, d),
                    );
                    if (!deeper) asserts = true;
                }
            });
            if (!asserts) continue;

            // A sentinel this function sets, asserted outside it, proves it ran.
            const sentinels = assignedNames(fn);
            // So does the result of the call it was handed to: a callback whose
            // caller returns its value cannot have been skipped if that value
            // is asserted. `const seen = withX(() => {...}); expect(seen)...`
            let host = null;
            walk(body, (n) => {
                if (n.type === "CallExpression" && n.arguments.includes(fn)) host = n;
            });
            if (host) {
                walk(body, (n) => {
                    if (
                        n.type === "VariableDeclarator" &&
                        n.init === host &&
                        n.id.type === "Identifier"
                    ) {
                        sentinels.add(n.id.name);
                    }
                });
            }
            let proven = false;
            walk(body, (n) => {
                if (proven) return;
                if (n.type !== "CallExpression" || n.callee.type !== "Identifier")
                    return;
                if (n.callee.name !== "expect") return;
                if (within(n, fn)) return;
                const t = text.slice(n.range[0], n.range[1]);
                for (const s of sentinels)
                    if (new RegExp(`\\b${s}\\b`).test(t)) proven = true;
            });
            if (proven) continue;

            const nameArg = node.arguments.find(
                (a) => a && (a.type === "Literal" || a.type === "TemplateLiteral"),
            );
            process.stdout.write(
                JSON.stringify({
                    file,
                    line: fn.loc.start.line,
                    method: fn._method ?? null,
                    test: nameArg
                        ? text
                              .slice(nameArg.range[0] + 1, nameArg.range[1] - 1)
                              .slice(0, 90)
                        : "(unnamed)",
                }) + "\n",
            );
        }
    });
}

File: tooling/architecture/js_unreached_assertions.py
#!/usr/bin/env python3
"""Ratchet assertions that a test can never be sure ran.

`js_vacuous_assertions.py` catches an assertion that cannot FAIL --
`expect(".o_x").toHaveCount(0)` where nothing declares `.o_x`. This one catches
an assertion that may never EXECUTE, which is the same defect one level down and
was gated by nothing.

The shape, from `t_custom_click.test.js` before `9e26c220cdf`:

    plop(ev) {
        expect(ev.currentTarget).toBe(document);      // runs only if the
    }                                                 // handler is invoked
    ...
    expect(".clickMe").toHaveText("text from props"); // runs always
    await contains(".clickMe").click();

One assertion executes, the test is green, and the assertion naming the
behaviour never ran. That test was the only coverage in `addons/web` for the
`.synthetic` modifier and it stayed green through the entire life of the defect
fixed in `800c9912f02`, where every synthetic handler after the first in a file
was silently dead.

MASKING IS THE DEFECT, NOT DEFERRAL. Hoot already fails a test that runs no
assertion at all, so a test whose ONLY assertions sit in a dead handler is caught
today. What is not caught is a dead handler masked by an assertion in the test's
own straight-line flow. This gate reports only the masked case; without that
condition it reports 112 sites in `addons/web`, most of them already covered.

Three things count as proof the deferred function ran, and none is reported:
`expect.assertions(n)` or the `verifySteps` protocol; a sentinel the function
assigns that is asserted outside it (`let clicked = false` ... `clicked = true`
... `expect(clicked).toBe(true)`, which is idiomatic here and correct); and the
result of the call the function was handed to, when that result is asserted.
"""

from __future__ import annotations

import argparse
import contextlib
import json
import subprocess
import sys
from dataclasses import asdict, dataclass
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _repo_root import find_odoo_root

ROOT = find_odoo_root(Path(__file__).resolve(), tool="js_unreached_assertions")
WEB_TESTS = ROOT / "addons" / "web" / "static" / "tests"

GOVERNED_ADDONS = ("web", "mail", "account", "stock")
DEFAULT_ADDON = "web"

ANALYZER = Path(__file__).with_suffix(".mjs")


def addon_tests(addon: str = DEFAULT_ADDON) -> Path:
    return (
        WEB_TESTS
        if addon == DEFAULT_ADDON
        else ROOT / "addons" / addon / "static" / "tests"
    )


@dataclass(frozen=True)
class Finding:
    file: str
    line: int
    test: str
    method: str | None

    def __str__(self) -> str:
        where = f"{self.method}()" if self.method else "callback"
        return f"  {self.file}:{self.line}  {where}  <- {self.test}"


def iter_test_files(tests: Path) -> list[Path]:
    if not tests.is_dir():
        return []
    return sorted(tests.rglob("*.test.js"))


def analyse(files: list[Path], tests: Path) -> list[Finding]:
    if not files:
        return []
    proc = subprocess.run(
        ["node", str(ANALYZER), *[str(f) for f in files]],
        capture_output=True,
        text=True,
        cwd=ROOT,
        check=False,
    )
    if proc.returncode != 0:
        raise RuntimeError(
            f"unreached-assertion analyzer failed: {proc.stderr.strip()}"
        )
    found = []
    for line in proc.stdout.splitlines():
        if not line.strip():
            continue
        raw = json.loads(line)
        rel = Path(raw["file"])
        with contextlib.suppress(ValueError):
            rel = rel.resolve().relative_to(tests.resolve())
        found.append(
            Finding(rel.as_posix(), raw["line"], raw["test"], raw.get("method"))
        )
    return sorted(found, key=lambda f: (f.file, f.line))


def measure(tests: Path = WEB_TESTS) -> list[Finding]:
    files = iter_test_files(tests)
    if not files:
        raise RuntimeError(
            f"no test files under {tests} -- the scan found nothing, which is "
            f"not the same as finding nothing wrong"
        )
    return analyse(files, tests)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--count", action="store_true", help="print the count only")
    parser.add_argument("--json", action="store_true", help="machine-readable")
    parser.add_argument("--top", type=int, default=20, help="0 for all")
    parser.add_argument(
        "--addon",
        default=DEFAULT_ADDON,
        choices=GOVERNED_ADDONS,
        help="which addon's static/tests to measure (default: web)",
    )
    args = parser.parse_args(argv)

    try:
        found = measure(tests=addon_tests(args.addon))
    except RuntimeError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    if args.count:
        print(len(found))
        return 0
    if args.json:
        print(json.dumps([asdict(f) for f in found], indent=2))
        return 0

    print(f"Assertions that may never execute ({args.addon}/static/tests)")
    print("=" * 72)
    shown = found if args.top == 0 else found[: args.top]
    for item in shown:
        print(item)
    if len(found) > len(shown):
        print(f"  ... and {len(found) - len(shown)} more (--top 0 for all)")
    print("-" * 72)
    print(f"{len(found)} masked assertion site(s)")
    print("\nRatchet this number:")
    suffix = "" if args.addon == DEFAULT_ADDON else f" --addon {args.addon}"
    name = "jsunreached" if args.addon == DEFAULT_ADDON else f"jsunreached_{args.addon}"
    print(
        f"  python tooling/architecture/js_unreached_assertions.py{suffix} --count \\"
    )
    print(f"      | xargs python tooling/ratchet/ratchet.py {name} --count")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
